keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,615 @@
+
+#### Proposed rearchitect
+
+With proposed rearchitecture we aim to:
+1. Reduce the potential scale issues when using lots of CronJob objects
+2. Reduce load on API server in such cases.
+3. Reduce memory usage
+
+##### Informers and Caches
+To reduce the need to list all Jobs and CronJobs frequently to reconcile, we propose to replace it with an Informers and WorkQueue based architecture. We would be sharing the [same informer cache](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L450) as the Job controller uses. 
+
+We will follow a controller structure similar to existing workload controllers and as [outlined in the guideline](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/controllers.md#rough-structure).
+```golang
+
+// 
+type CronJobController struct {
+	kubeClient 			clientset.Interface
+	queue	   			workqueue.RateLimitingInterface
+	recorder   			record.EventRecorder
+	
+	jobControl 			jobControlInterface
+	cronJobControl  		cronJobControlInterface
+	
+	jobLister  			batchv1listers.JobLister
+	cronJobLister 			CronJobLister
+	
+	jobListerSynced 		cache.InformerSynced
+	cronJobListerSynced		cache.InformerSynced
+}
+
+// CronJobControllerConfiguration contains controller config
+type CronJobControllerConfiguration struct {
+	// concurrentCronJobSyncs is the number of cronjob objects that are
+	// allowed to sync concurrently. Larger number = more responsive
+	// but more CPU (and network) load.
+	ConcurrentCronJobSyncs int32
+}
+
+
+// Controller code
+
+func (cjc *CronJobController)  Run() {
+	...
+	if !cache.WaitForNamedCacheSync("cronjob", stopCh, cjc.jobListerSynced, cjc.cronJobListerSynced) {
+		return
+	}
+    // start multiple workers
+    for i := 0; i < workers; i++ {
+        go wait.Until(cjc.worker, time.Second, stopCh)
+    }
+...
+}
+
+func (cjc *CronJobController) func worker() {
+    for cjc.processNextWorkItem() {}
+}
+
+func (cjc *CronJobController) func processNextWorkItem() {
+	key, quit := cjc.queue.Get()
+	if quit {
+		return false
+	}
+	defer cjc.queue.Done(key)
+	if err := cjc.sync(key.(string)); err != nil {
+		utilruntime.HandleError(fmt.Errorf("Error syncing CronJobController %v, requeuing: %v", key.(string), err))
+		cjc.queue.AddRateLimited(key)
+	} else {
+		cjc.queue.Forget(key)
+	}
+	return true
+}
+
+func (cjc *CronJobController)  sync(cronJobKey) {
+	childrenJobs := cjc.jobLister.GetJobsForCronJob(cronJobKey)
+	// reuse/refactor the existing syncOne function
+}
+```
+
+##### Multiple workers
+We also propose to have multiple workers controller by a flag similar to [statefulset controller](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/apps.go#L65). The default would be set to 5 similar to [statefulset](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/config/v1alpha1/defaults.go#L34)
+
+##### Handling Cron aspect
```

Comment:
There is a critical problem with this current implementation that I would like to call out.

When determining if a cron should be scheduled, `getMissedSchedules()` is used to return the most recent missed schedule time (and a count of missed start times, in the form of extra items in the returned slice). https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cronjob/utils.go#L92

This code uses a 3rd party module that has a `Next()` function for schedules, but not a `Previous()`. Rather than looking back at `now.Previous()`, the code repeatedly calls `Next()`, until hitting reaching the current time... or hitting a hard cap. I have explored 3 ways to remove the hard cap, ranked in decreasing order of appeal as I understand them:

1. Implement a Previous function, by working around the 3rd party library, or by outright replacing it.
2. Use a binary search with the `Next()` function, to avoid iterating over every start time in the checked time window. For any subwindow, if there is a is start time in the second half of the window, the most recent missed start time is at *least* at that time.
3. Remove the concept of "which start time" the job is launched on behalf of. If we only need to know that the Cron Job is due for a run (and is within its startIngDeadlineSeconds), we can avoid this pattern. This seems the least ideal due to the decrease in clear status.

We are implementing option 2 at Lyft (as it's far easier to test and have fast confidence in than option 1). 100 missed start times is nowhere near long enough to support our production Cron users, given the myriad of ways that code or the CronJob/Kubernetes machinery can break.

Reply:
@vllry  - Thanks for this feedback. Are you using a custom version of CronJob or a separate controller ? Is that public ? 

We are interested in the myriad ways the machinery can break.

Reply:
We are patching the upstream controller (I can open an open source PR soon if you're interested, but it's not yet deployed/proven internally). We considered writing a new controller but are hoping this effort moves forward soon.

We have ~250 CronJobs at present, which entails hitting (a) a lot of infrequent edge cases, and (b) being extremely susceptible to the performance issues outlined in the KEP and my specific comment.